docs: fees tutorial (#13357)
AztecProtocol/aztec-packages: docs/docs/developers/tutorials/codealong/first_fees.md

**The PXE**
The PXE is a client-side key manager, private contract storage, and Private eXecution Environment for private transactions. It interfaces with the Aztec network, and can be used via its json RPC client. A PXE is a core part of an Aztec wallet and Sandbox, but can be decoupled and run independently.

**An Aztec node**
An Aztec node is a prover/sequencer that is part of a decentralised Aztec network. The Aztec testnet rolls up to Ethereum Sepolia.

**The Aztec Sandbox**
The Aztec Sandbox runs a local environment for rapid development, it includes: an Ethereum node, an Aztec node, and PXE.

## Connect to the network

We'll go through both the `aztec-wallet` cli tool and the `aztec.js` library, to a local Aztec network (in the sandbox) and the Aztec testnet.

### Tools

**TODO: as snippet**
To install the required tools see [Getting Started](../../getting_started.md) or just run `bash -i <(curl -s https://install.aztec.network)`

Test the CLI tool with: `aztec-wallet --version`

By default the sandbox runs everything including a pxe, but for this tutorial and more realistically when using testnet, we will be using PXEs in the wallet, not the sandbox.

> Start the sandbox (L1, L2, but not the PXE) via: `aztec start --sandbox.noPXE`

Reviewer: this command doesn't work.

I tried `aztec start --sandbox -no-pxe` like we suggest in aztec-starter. It starts the sandbox, but also starts the pxe. It's fine if the PXE starts, we can still start another one and just use that one for a more realistic setup.

```bash
aztec-wallet create-account --register-only -a main
```

The equivalent in Aztec.js:

```javascript

```

Your PXE now has keys for an account that can be deployed on Aztec network(s).

## Paying for an account deployment transaction

To make transactions on the network, your account contract will need to specify a payment method of the enshrined asset, "Fee Juice".

TODO: types of payment methods, one-liner each
- FPCs
  - Sponsored
  - Public / Private
- bridge and claim
- Fee juice

### Sponsored FPC

To bootstrap first use, a sponsored fee paying contract (the canonical sponsored FPC) can be used to deploy your first account.

**TODO: as snippet**
A fee paying contract (FPC) effectively implements fee abstraction. It is a contract that pays for transactions of other accounts, when its own custom criteria is met.
In the case of the canonical sponsored FPC, the only criteria is an upper bound on how much it sponsors an account's transactions. This will be enough to at least deploy an account.

The PXE can be queried for the canonical sponsored FPC address, and then specified as the payment method.

Via the CLI:

```bash

> FPC_ADDRESS=`aztec-wallet --get-canonical-sponsored-fpc-address`

Reviewer: Is this the actual flag?

I see:
```bash
josh in ~ λ aztec-wallet --get-canonical-sponsored-fpc-address
error: unknown option '--get-canonical-sponsored-fpc-address'
```

Author: It as only merged 1-2 days ago, and there hasn't been a release.

---
title: All about fees

> sidebar_position: 0

Reviewer: Do you think the fees tutorial should be the first in the list? I think it might be better after "Contract Tutorials" or after "Aztec.js Tutorials". Just thinking it would be better to introduce to people to the meatier topics first.

Reviewer 2: sorry coming to the conversation late - our tutorials are very linear so i dont think it makes sense for here. i'd pop it between contract & aztecjs tutorials for now. since we now have 2 tutorials that dont sit nicely in a category, i think we should change this sidebar very slightly (and be v careful to add redirects)

- writing contracts
    - counter
    - private voting
    - crowdfunding
    - account
    - nft
- full stack apps
    - transferring tokens with aztecjs
     - dapp
    - token bridge
    - uniswap
- fees (could also sit in fullstack apps but youve written it from setup onwards so i think its nice by itself)
- cli wallet

wdyt? i can do this in another pr

Reviewer 2: @jzaki

Reviewer: i update this

Author: It is intended to be consumed linearly where devs can choose between a few paths based on their leaning/level:
- CLI only (view aztec.js for reference)
- .js only (view CLI refs for reference)
- both in parallel

Finally, it also serves as a useful reference to come back to.

This is all by design, but if you wish to tuck it away somewhere, I'll leave it for a future pr.

**An Aztec node**
An Aztec node is a prover/sequencer that is part of a decentralised Aztec network. The Aztec testnet rolls up to Ethereum Sepolia.

**The Aztec Sandbox**
The Aztec Sandbox runs a local environment for rapid development, it includes: an Ethereum node, an Aztec node, and PXE.

## Connect to the network

We'll go through both the `aztec-wallet` cli wallet and the `aztec.js` library, to a local Aztec network (in the sandbox) and the Aztec testnet.

### Tools

**TODO: as snippet**
To use Aztec's suite of tools you'll need to:
 - [Get docker](https://docs.docker.com/get-started/get-docker/)
 - Run `bash -i <(curl -s https://install.aztec.network)`

Test the cli wallet with: `aztec-wallet --version`

By default the sandbox runs everything including a pxe. For this tutorial, and more realistically when using testnet, we will be using PXEs client side. Eg as part of a wallet.

Start the sandbox (L1, L2, but not the PXE) via: `NO_PXE=true aztec start --sandbox`

### Specifying the network URL in commands

When using `aztec-wallet`, the Aztec network node to connect to can be specified in each relevant command:

```bash
aztec-wallet --node-url <string> ...
```

> The string currently defaults to the the local sandbox: `http://host.docker.internal:8080`, or can be a bootnode url.

Reviewer: does it default to this or localhost:8080? or is it functionally the same thing?

Author: I saw the host in one file, and localhost in another, so it would be the same if the container is exposing it locally. (Or localhost if connecting things within the docker container).
Just reading an internal thread about sandbox/testnet releases, and I'm not sure if this can correctly be pointed to testnet now (might just be the default in the build?).

Will get clarity once it settles and update once we can test it.

More conveniently, this can be specified via the environment variable: `AZTEC_NODE_URL`.

The equivalent using aztec.js - create a PXE pointing to the desired url:

```javascript
import { createPXEClient, waitForPXE, PXE } from "@aztec/aztec.js";
async function main() {
  const { PXE_URL = 'http://localhost:8080' } = process.env; // from env var
  const pxe = await createPXEClient(PXE_URL);
  await waitForPXE(pxe);
  // use pxe...
}
```

## Create Account Contract in a PXE

**TODO: as snippet**

An account on Aztec is a smart contract that specifies a method of authentication and a method of payment, allowing it to be used by the protocol to perform a transaction.

> For convenience, Aztec has implemented an account contract that authenticates transactions using Schnorr signatures. The contract class for a Schnorr account is pre-registered on the network to bootstrap first use. Ordinarily for a contract to be deployed, its class would have to be register with the network first.

Reviewer: ```suggestion
For convenience, Aztec Labs has implemented an account contract that authenticates transactions using Schnorr signatures. The contract class for a Schnorr account is pre-registered on the sandbox and testnet networks to bootstrap first use. Ordinarily for a contract to be deployed, its class would have to be register with the network first.
```

Author: Updated `on Aztec networks (eg sandbox, testnet)`

When a PXE creates a Schnorr account, there are three key things that occur:
 - generation of keys privately
 - registeration of a Schnorr account locally (effectively an instance of the Schnorr account class)
   - at this stage the address of the account is known, even though it is not yet deployed
 - initialization and deployment of the account contract to the network

**Create the keys and account in the PXE**

> Lets first create the account but only register the instance in the pxe. Its address will be calculated, and in a later step we can deploy it counterfactually to the network.

Reviewer: ```suggestion
Lets first create the account but only register the instance in the PXE. Its address will be calculated, and in a later step we can deploy it to the network.
```
I think the term "counterfactual" may be more confusing than helpful here.

```bash
aztec-wallet create-account --register-only -a main
```

The `-a main` sets this new account's alias to, "main". Use `aztec-wallet get-alias` (with no params) to see all aliases.

The equivalent using aztec.js - create a random account locally (we will deploy it in the next step):

```javascript
import { Fr } from "@aztec/aztec.js";
import { getSchnorrAccount } from '@aztec/accounts/schnorr';
async function main() {

>   let secretKey = Fr.random(); let salt = Fr.random();

Reviewer: It's not clear in this snippet that the code from the above snippet is required for this to work

Author: Added a code comment

  let schnorrAccount = await getSchnorrAccount(pxe, secretKey, deriveSigningKey(secretKey), salt);
}
```

Your PXE now has keys for an account that can be deployed on Aztec network(s).

## Paying for an account deployment transaction

To make transactions on the network, your account contract will need to specify a payment method of the enshrined asset, "Fee Juice".

### Sponsored FPC

To bootstrap first use, a sponsored fee paying contract (the canonical sponsored FPC) can be used to deploy your first account.

**TODO: as snippet**
A fee paying contract (FPC) effectively implements fee abstraction. It is a contract that pays for transactions of other accounts, when its own custom criteria is met.
In the case of the canonical sponsored FPC, the only criteria is an upper bound on how much it sponsors an account's transactions. This will be enough to at least deploy an account.

The PXE can be queried for the canonical sponsored FPC address, and then specified as the payment method.

Via the CLI:

The alias set earlier can be confirmed using: `aztec-wallet get-alias accounts:main`, this is specified here in `--from main`.

```bash

> SPONSORED_FPC_ADDRESS=`aztec-wallet --get-canonical-sponsored-fpc-address`

Reviewer: ```suggestion
aztec-wallet register-contract contracts:sponsoredfpc SponsoredFPC --from main
```
The address is at alias `contracts:sponsoredfpc`. You need to register it in the `main` account PXE before you can use it

Author: The alias was not set for me, but updated.

> aztec-wallet deploy-account --from main --payment method=fpc-sponsored,fpc=$SPONSORED_FPC_ADDRESS

Reviewer: ```suggestion
aztec-wallet deploy-account --from main --payment method=fpc-sponsored,fpc=contracts:sponsoredfpc
```
Once its registered you can deploy with this command

```

The equivalent using aztec.js - get sponsored fpc address and use payment method:

```javascript
import { SponsoredFeePaymentMethod } from "@aztec/aztec.js/fee/testing";

> import { getDeployedSponsoredFPCAddress } from "../src/utils/sponsored_fpc.js"; // currently manually brought into project (see aztec-starter repo)

Reviewer: I think you want to get `getSponsoredFPCInstance` like [here](https://github.com/AztecProtocol/aztec-starter/pull/80/files#diff-36d808495de81c0ef334263da01fb16ea4380b9d54c9ca49c3f445eb4f382650R74-R76). I think what you have will work in the sandbox, but not if a PXE is running separately.

Author: Updated.

  //...
  const sponseredFPCAddress = await getDeployedSponsoredFPCAddress(pxe);
  const sponsoredPaymentMethod = new SponsoredFeePaymentMethod(sponseredFPCAddress); // create payment method
  let tx = await schnorrAccount.deploy({ fee: { paymentMethod: sponsoredPaymentMethod }}).wait();
  let wallet = await schnorrAccount.getWallet(); let address = wallet.getAddress();

```

:::note Payment: Sponsored Fee Paying Contract
Options for payment via the sponsored fpc that can be used in multiple commands:

> CLI: `--payment method=fpc-sponsored,fpc=$SPONSORED_FPC_ADDRESS`

Reviewer: ```suggestion
CLI: `--payment method=fpc-sponsored,fpc=contracts:sponsoredfpc`
```

Author: Given this is the generic line in isolation, and the internal thread, I think the original "address" makes more sense than an alias that we haven't shown being defined.

.js: `{ fee: { paymentMethod: new SponsoredFeePaymentMethod(sponseredFPCAddress) }}`
:::

**Congratulations! You have successfully created an account on Aztec!**

This contract now exists in the sandbox network, or on testnet if you specified a node url.

### Fee Juice

Apart from the FPC payment methods, the default method for paying for transactions is via fee juice direct from the sender of the tx.

> For the special case of deploying an account, another funded account can be specified to pay for the deployment transaction.

Reviewer: ```suggestion
For the special case of deploying an account, another funded account can be specified to pay for the deployment transaction, or you can use fee juice that has been bridged to the address of the to-be-deployed account.
```

:::

## Summary of fee payment options

The two key ways of paying for transactions: fee juice from an account or via an FPC. Both of which are contracts on the aztec network.

### Fee juice from an account (default)

- from the sender of a transaction (default)
  - most common if making transactions from a funded account
  - difficult to deploy without funds already
- specifying a different fee-payer account (new account creation/deployment only)
  - using a funded account to deploy another account, or rapid testing on the sandbox
  - need to already have a funded account
- claiming fee juice from already-bridged L1 asset and immediately using
  - great for bootstrapping an account on Aztec
  - need to already have aztec fee asset on an L1 account

### Fee juice from a fee paying contract

- via public/private payment
  - can pay in other assets, and not hold fee juice
  - can privately spend alternative assets, to make private transactions
- sponsored payment
  - enables new users to bootstrap their first account

### Example payment options

```bash

> aztec-wallet <command> --payment method=fee_juice # default fee juice from sender of transaction

Reviewer: might be easier to read with the comments between the lines

aztec-wallet <create/deploy command> --payment method=fee_juice,feePayer=$FEE_PAYER_ADDRESS # fee juice from another account (create/deploy txs only)
aztec-wallet <command> --payment method=fee_juice,claim # claim bridged fee juice and pay
aztec-wallet <command> --payment method=fpc-sponsored,fpc=$SPONSORED_FPC_ADDRESS # sponsored fpc
aztec-wallet <command> --payment method=fpc-public,fpc=$FPC_ADDRESS,asset=$ASSET_ADDRESS # asset for fee juice via fpc public
aztec-wallet <command> --payment method=fpc-private,fpc=$FPC_ADDRESS,asset=$ASSET_ADDRESS # asset for fee juice via fpc private

```

```javascript

> `command({ fee: { paymentMethod: new FeeJuicePaymentMethod(<sender address>) }})` // default fee juice from sender of transaction

Reviewer: I don't think you need the ` since its already wrapped in a code block. and it might be easier to read these code snippets with some line breaks and putting the comments in between the lines.

Author: Updated.

`command({ fee: { paymentMethod: new FeeJuicePaymentMethod(<fee payer address>) }})` // different from sender (create/deploy txs only)
`command({ fee: { paymentMethod: new FeeJuicePaymentMethodWithClaim(newWallet, claim) }})` // claim bridged fee juice and pay
`command({ fee: { paymentMethod: new SponsoredFeePaymentMethod(sponseredFPCAddress) }})` // sponsored fpc
`command({ fee: { paymentMethod: new PublicFeePaymentMethod(fpc.address, newWallet) }})` // asset for fee juice via fpc public
`command({ fee: { paymentMethod: new PrivateFeePaymentMethod(fpc.address, newWallet) }})` // asset for fee juice via fpc private

```

:::warning
Lists above are provided as a convenient reference and are not automatically maintained.
Please refer to the snippets in the sections above, and report any discrepancies.

:::

### Tabulated payment methods and options

`aztec-wallet` CLI params:

method\options|`feePayer`|`asset`|`fpc`|`claim`
-|-|-|-|-

> `fee_juice`|create/deploy account only|NA|NA|if bridged

Author: The table refers to the CLI options, although I can see how the underlying asset would be fee juice. I'll try make it clearer

> `fpc-public`|NA|asset address|contract address|NA

Reviewer: ```suggestion
`fpc-public`|FPC|asset address|contract address|NA
```

> `fpc-private`|NA|asset address|contract address|NA

Reviewer: ```suggestion
`fpc-private`|FPC|asset address|contract address|NA
```

Author: as above

---
title: All about fees
sidebar_position: 0
tags: [fees, accounts, transactions, cli, contracts]
---

> The Aztec network is a privacy preserving layer 2 secured by Ethereum, and for the consensus mechanism to work, Aztec makes use of an asset to pay for transactions called, "Fee juice".

Reviewer: ```suggestion
The Aztec network is a privacy preserving layer 2 secured by Ethereum, and for the consensus mechanism to work, Aztec makes use of an asset to pay for transactions called, "Fee juice".  Fee juice is non- transferable on Aztec, which creates some user experience hurdles. To get around this, Aztec has fee abstraction, which means accounts can pay transaction fees on behalf of other accounts. This tutorial will demonstrate some of the ways to take advantage of fee abstraction and pay transaction fees on Aztec (e.g. Fee Payment Contracts).
```

Author: I don't want to deter readers in the first paragraph, but I'll amend.

```bash
aztec-wallet create-account --register-only -a main
```

The `-a main` sets this new account's alias to, "main". Use `aztec-wallet get-alias` (with no params) to see all aliases.

The equivalent using aztec.js - create a random account locally (we will deploy it in the next step):

```javascript
import { Fr } from "@aztec/aztec.js";
import { getSchnorrAccount } from '@aztec/accounts/schnorr';
  //... building upon the previous section
  let secretKey = Fr.random(); let salt = Fr.random();
  let schnorrAccount = await getSchnorrAccount(pxe, secretKey, deriveSigningKey(secretKey), salt);
```

Your PXE now has keys for an account that can be deployed on Aztec network(s).

## Payment option overview

To get a quick overview of the payment options currently supported...

**Sponsored Fee Paying Contract (Sponsored FPC)**
A small amount of funds from a limited faucet-like contract. Useful for bootstraping first tx(s), eg deloying an account

**Fee juice**
- Claimed from the fee asset bridged from L1
- From a funded account sending a transaction (default)

> - From a different funded account to deploy an account, (not possible with other tx types, only deployment)

Author: The special case of one account contract paying for the deployment of another account contract, which in the cli commands was only possible.
See `aztec-wallet help deploy-account` specifies `feePayer`, whereas `aztec-wallet help send` does not.

Remember, we will go into these in the following sections.

## Paying for an account deployment transaction

To make transactions on the network, your account contract will need to specify a payment method of the enshrined asset, "Fee Juice".

### Sponsored FPC

To bootstrap first use, a sponsored fee paying contract (the canonical sponsored FPC) can be used to deploy your first account.

**TODO: as snippet**
A fee paying contract (FPC) effectively implements fee abstraction. It is a contract that pays for transactions of other accounts, when its own custom criteria is met.
In the case of the canonical sponsored FPC, the only criteria is an upper bound on how much it sponsors an account's transactions. This will be enough to at least deploy an account.

The PXE can be queried for the canonical sponsored FPC address, and then specified as the payment method.

Via the CLI:

The alias set earlier can be confirmed using: `aztec-wallet get-alias accounts:main`, this is specified here in `--from main`.

```bash
SPONSORED_FPC_ADDRESS=<address>
aztec-wallet register-contract SPONSORED_FPC_ADDRESS SponsoredFPC --from main # Need to specify account that wishes to register the contract

> aztec-wallet deploy-account --from main --payment method=fpc-sponsored,fpc=contracts:sponsoredfpc

Reviewer: ```suggestion
aztec-wallet deploy-account --from main --payment method=fpc-sponsored,fpc=$SPONSORED_FPC_ADDRESS
```

Author: This was aliased in the previous command, but I'll remove the alias to use the env var.

```

The equivalent using aztec.js - get sponsored fpc address (helper functions [here](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec/src/sandbox/sponsored_fpc.ts
)) and use payment method:

```javascript
import { getSponsoredFPCInstance } from "../src/utils/sponsored_fpc.js"; // helper functions linked above
import { SponsoredFPCContract } from "@aztec/noir-contracts.js/SponsoredFPC";
import { SponsoredFeePaymentMethod } from "@aztec/aztec.js/fee/testing";
  //... building upon the previous section
  const sponseredFPC = await getSponsoredFPCInstance(); // get address of pre-deployed Sponsored FPC contract
  await pxe.registerContract({instance: sponseredFPC, artifact: SponsoredFPCContract.artifact}); // register the Sponsord FPC contract class with the pxe
  const sponsoredPaymentMethod = new SponsoredFeePaymentMethod(sponseredFPC.address); // create payment method
  let tx = await schnorrAccount.deploy({ fee: { paymentMethod: sponsoredPaymentMethod }}).wait();
  let wallet = await schnorrAccount.getWallet(); let address = wallet.getAddress();

```

> :::note Payment: Sponsored Fee Paying Contract

Author: These payment info boxes appear more than once, and are easier to see when scrolling quickly between them. 

I've added some spacing between the code lines (for each of the info boxes)

Options for payment via the sponsored fpc that can be used in multiple commands:
- CLI: `--payment method=fpc-sponsored,fpc=$SPONSORED_FPC_ADDRESS`
- .js: `{ fee: { paymentMethod: new SponsoredFeePaymentMethod(sponseredFPCAddress) }}`
:::

**Congratulations! You have successfully created an account on Aztec!**

This contract now exists in the sandbox network, or on testnet if you specified a node url.

### Fee Juice

> Apart from the FPC payment methods, the default method for paying for transactions is via fee juice direct from the sender of the tx.

Reviewer: ```suggestion
Apart from the FPC payment methods, the default method for paying for transactions is via fee juice direct from the sender of the transaction.
```

> When specifying fee juice, and only in the special case of deploying an account, a different funded account may be specified to pay for the deployment transaction of an unfunded account.

Reviewer: I just tried by deploying the counter and calling it with:
```bash
aztec-wallet send increment -ca counter --args accounts:test0 accounts:test0 --from main method=fee_juice,feePayer=test0
```
and it works
```suggestion
When specifying fee juice, a different funded account may be specified to pay for the deployment transaction of an unfunded account.
```

Another way for an accounts deployment to be funded is via claiming bridged fee juice (in a later section).

#### Sandbox pre-funded test accounts

The sandbox starts with 3 test accounts, providing another way to bootstrap initial testing with accounts. To add these to the PXE...

For the CLI:
```bash
aztec-wallet import-test-accounts
```

Confirm with: `aztec-wallet get-alias`, to see all aliases

The equivalent using aztec.js - get the test accounts
```javascript
  import { getInitialTestAccountsWallets } from "@aztec/accounts/testing";
  //...

>     testWallets = await getInitialTestAccountsWallets(pxe);

Reviewer: ```suggestion
    const testWallets = await getInitialTestAccountsWallets(pxe);
```

>     testWallets = await getInitialTestAccountsWallets(pxe);

Reviewer: And this only works when running the sandbox with the PXE attached. You can't use it when running an independent PXE.

Author: This is in the section `#### Sandbox pre-funded test accounts`, but are you saying that if they're running the sandbox and pxe separately, they can't do this?

```

Note: The test account addresses can also be seen in the sandbox logs when it starts up.

**Create and deploy a new account, paid for by another account**

For the sandbox only, you can use the test accounts provided at aliases: `test0`, `test1`, `test2`

```bash
aztec-wallet create-account -a alice --payment method=fee_juice,feePayer=test0
```

:::warning

> Specifying the `feePayer` is only an option for commands that deploy an account.

Reviewer: ```suggestion
```
not true

Author: feePayer in cli options is used in [create account](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/cli-wallet/src/cmds/index.ts#L104) and [deploy account](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/cli-wallet/src/cmds/index.ts#L158) only.
Otherwise param to inner function is false and does not use `feePayer`

```javascript
import { FeeJuicePaymentMethod } from "@aztec/aztec.js";
  //... building upon the previous section
  const useFeeJuice = new FeeJuicePaymentMethod(testWallets[0].getAddress())
  await schnorrAccount.deploy({ fee: { paymentMethod: useFeeJuice }}).wait()
```

:::note Payment: Sponsored Fee Paying Contract
Options for explicitly stating fee_juice from the sender which is the default payment method:
- CLI: `--payment method=fee_juice` (default)
- .js: `{ fee: { paymentMethod: new FeeJuicePaymentMethod(<fee payer address>) }`
:::

### Bridging Fee Juice

The sandbox allows free-minting on it's L1 to be bridged and claimed on its Aztec node.
For testnet you'll first need an L1 account with the fee asset to send the fees to the Aztec bridge.

We'll register a new account `accBFJ` and bridge fee-juice to it.

```bash
aztec-wallet create-account -a accBFJ --register-only
aztec-wallet bridge-fee-juice 1000000000000000000 accBFJ --mint --no-wait
```

You'll have to wait for two blocks to pass for bridged fee juice to be ready on Aztec. For the sandbox you can do this by putting through two arbitrary transactions. Eg:
```bash

> aztec-wallet deploy counter_contract@Counter --init initialize --args 0 accounts:test0 --from test0 -a counter

Reviewer: ```suggestion
aztec-wallet deploy Counter --init initialize --args 0 accounts:test0 --from test0 -a counter
```

aztec-wallet send increment -ca counter --args accounts:test0 accounts:test0 --from test0
```

Now the funded account can deploy itself with the bridged fees, claiming the bridged fee juice and deploying the contract in one transaction:

```bash
aztec-wallet deploy-account --from accBFJ --payment method=fee_juice,claim
```

The equivalent using aztec.js - bridge fee juice, (pass two txs), create and use payment method:

(See [here](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/cli-wallet/src/cmds/bridge_fee_juice.ts#L32) to initialise a fee juice portal manager)

```javascript
import { FeeJuicePaymentMethod, PrivateFeePaymentMethod, PublicFeePaymentMethod } from "@aztec/aztec.js";
  const newWallet = await schnorrAccount.getWallet()
  const feeJuiceReceipient = schnorrAccount.getAddress()

>   const claim = await feeJuicePortalManager.bridgeTokensPublic(feeJuiceReceipient, 1000000000000000000n, true);

Reviewer: How do you set up `feeJuicePortalManager`? This is the first time you use it.

Reviewer: you can see how its done here https://github.com/AztecProtocol/aztec-starter/blob/dd2b385ab92758343e8ff9a7b1dcca29e7c8c8c0/scripts/fees.ts#L69

Author: Above this I link to code that shows an implementation using @aztec/etherem, I can instead bring that code in here, but it makes the code snippet much larger than the bite-size chunks I had aimed for.

The code linked to is from aztec-wallet (bridgeL1FeeJuice command in bridge_fee_juice.ts), and was deliberately chosen to communicate the .ts inside `aztec-wallet` is a source for other things they may need to do.

  // ...(wait or perform two txs)
  const claimAndPay = new FeeJuicePaymentMethodWithClaim(newWallet, claim)
  await schnorrAccount.deploy({ fee: { paymentMethod: claimAndPay }}).wait()

```

:::note  Payment: Fee Juice with claim from bridge
Options for claim+pay with bridged funds that can be used in multiple commands:

> - CLI: `--from <sender address> --payment method=fee_juice,claim`

Reviewer: I think it would be good to take the code snippets out of the :::note block here too. It would look better outside and in the context of a full command / JS call

Author: Presented better as above. Note blocks are for quick scrolling reference. Can change if the tweak doesn't make sense still

- .js: `{ fee: { paymentMethod: new FeeJuicePaymentMethodWithClaim(newWallet, claim) }}`
:::

### Fee Paying Contract payment (public/private) - Advanced

> 

Reviewer: Maybe link them to this [script](https://github.com/AztecProtocol/aztec-starter/blob/dd2b385ab92758343e8ff9a7b1dcca29e7c8c8c0/scripts/fees.ts#L97) for a reference on how to do it?

Setting up your own FPC and authorising the asset is an involved process outside the scope of this tutorial.
Below we will look at the syntax for understanding.

First register the FPC address in your pxe. In reality this might be an application funding users' transactions via their token.

```bash
aztec-wallet register-contract $FPC_ADDRESS FPCContract --from main
aztec-wallet <command> --payment method=fpc-public,fpc=$FPC_ADDRESS,asset=$ASSET_ADDRESS
```

The second line can be any command that takes a `--payment` parameter. See `aztec-wallet --help` and the help of corresponding commands to check.

> Notice the specified address of both the fpc contract, and the asset address.

Reviewer: ```suggestion
Notice the specified address of both the FPC contract, and the asset address.
```

The equivalent using aztec.js:

```javascript
import { FPCContract } from "@aztec/noir-contracts.js/FPC";
  //...

>   const fpc = await FPCContract.deploy(wallets[0], bananaCoin.address, wallets[0].getAddress()).send().deployed();

Reviewer: Mention that banana coin is a token that the FPC is accepting for payment

  // ... (setup token authorisation, and fpc balance here)
  const publicFee = new PublicFeePaymentMethod(fpc.address, newWallet)

```

:::note  Payment: Fee Paying Contract (public/private)
Options for payment via fpc that can be used in multiple commands:

> CLI: `--payment method=fpc-public,fpc=$FPC_ADDRESS,asset=$ASSET_ADDRESS` (`fpc-private` for private)

Reviewer: take the code snippets out of this admonition as well

.js: `{ fee: { paymentMethod: new PublicFeePaymentMethod(fpc.address, newWallet) }` (`PrivateFeePaymentMethod` for private)
:::

:::info Private fee payment

> Note: using a private fpc method is the only way for transactions to be paid for privately.

Reviewer: ```suggestion
Note: using a private FPC method is the only way for transactions to be paid for privately.
```

> Public and private refer to how the fpc will claim its tokens from the sender to then (publicly) spend fee juice. The visibility of the function being called is unchanged.

Reviewer: ```suggestion
Public and private refer to how the FPC will claim its tokens from the sender to then (publicly) spend fee juice. The visibility of the function being called is unchanged.
```

:::

## Summary of fee payment options

The two key ways of paying for transactions: fee juice from an account or via an FPC. Both of which are contracts on the aztec network.

### Fee juice from an account (default)

- from the sender of a transaction (default)

>   - most common if making transactions from a funded account

Reviewer: ```suggestion
  - if making transactions from a fee juice funded account
```
I wouldn't say we have common conventions yet :slightly_smiling_face:

>   - difficult to deploy without funds already

Reviewer: ```suggestion
  - requires bridging the fee asset from L1 to L2
```
not difficult, just requires an additional step

Author: And the preceding steps to acquire fee juice. This is making the point for devs that their users will need to acquire fee asset on L1 first, which requires an Ethereum wallet...
The best ux a dev can offer is some form of gating that funds their users on Aztec.

Author: Amended

> - specifying a different fee-payer account (new account creation/deployment only)

Reviewer: ```suggestion
- specifying a different fee-payer account
```

Author: My understanding with how it works under the hood, was that a transaction has a "sender" account, and it is only the case of account creation/deployment that another account can be the sender.
This is different to other contracts as the sender.

Reviewer: I am going off of what is possible via the commands. I mentioned above that specifying a different `--from` address and `--feePayer` address works with the CLI

Reviewer: After discussing with @jzaki earlier, I may have tested this incorrectly. Worth testing again.

  - using a funded account to deploy another account, or rapid testing on the sandbox
  - need to already have a funded account
- claiming fee juice from already-bridged L1 asset and immediately using
  - great for bootstrapping an account on Aztec
  - need to already have aztec fee asset on an L1 account

> ### Fee juice from a fee paying contract

Reviewer: ```suggestion
### Fee juice from a Fee Paying Contract
```

- via public/private payment
  - can pay in other assets, and not hold fee juice
  - can privately spend alternative assets, to make private transactions
- sponsored payment
  - enables new users to bootstrap their first account

### Example payment options

```bash
# default fee juice from sender of transaction
aztec-wallet <command> --payment method=fee_juice

> # fee juice from another account (create/deploy txs only)

Reviewer: ```suggestion
# fee juice from another account
```

Author: As before, specifying a feePayer is unique to create/deploy transactions

command({ fee: { paymentMethod: new FeeJuicePaymentMethod(<sender address>) }})

// fee juice from different sender (create/deploy txs only)
command({ fee: { paymentMethod: new FeeJuicePaymentMethod(<fee payer address>) }})

// claim bridged fee juice and pay
command({ fee: { paymentMethod: new FeeJuicePaymentMethodWithClaim(newWallet, claim) }})

// sponsored fpc
command({ fee: { paymentMethod: new SponsoredFeePaymentMethod(sponseredFPCAddress) }})

// asset for fee juice via fpc public
command({ fee: { paymentMethod: new PublicFeePaymentMethod(fpc.address, newWallet) }})

// asset for fee juice via fpc private
command({ fee: { paymentMethod: new PrivateFeePaymentMethod(fpc.address, newWallet) }})
```

:::warning
Lists above are provided as a convenient reference and are not automatically maintained.
Please refer to the snippets in the sections above, and report any discrepancies.

:::

### Tabulated payment methods and options

`aztec-wallet` method and option parameters.
- The rows correspond to the method, eg `--payment method=fee_juice`
- The columns are the options, eg `--payment method=fpc-sponsored,fpc=<address>`

method\options|`feePayer`|`asset`|`fpc`|`claim`
-|-|-|-|-

> `fee_juice`|create/deploy account only|NA|NA|if bridged

Reviewer: ```suggestion
`fee_juice`|create/deploy account only|fee juice|NA|if bridged
```

> `fpc-public`|NA|asset address|contract address|NA

Reviewer: ```suggestion
`fpc-public`|NA|FPC accepted asset|contract address|NA
```

> `fpc-private`|NA|asset address|contract address|NA

Reviewer: ```suggestion
`fpc-private`|NA|FPC accepted asset|contract address|NA
```
